reprovide/reprovide.go

```diff
@@ -0,0 +1,54 @@
+package reprovide
+
+import (
+	"time"
+
+	"github.com/jbenet/go-ipfs/Godeps/_workspace/src/code.google.com/p/go.net/context"
+	blocks "github.com/jbenet/go-ipfs/blocks/blockstore"
+	routing "github.com/jbenet/go-ipfs/routing"
+	"github.com/jbenet/go-ipfs/util/eventlog"
+)
+
+var log = eventlog.Logger("reprovider")
+
+type Reprovider struct {
+	// The routing system to provide values through
+	rsys routing.IpfsRouting
+
+	// The backing store for blocks to be provided
+	bstore blocks.Blockstore
+}
+
+func NewReprovider(rsys routing.IpfsRouting, bstore blocks.Blockstore) *Reprovider {
+	return &Reprovider{
+		rsys:   rsys,
+		bstore: bstore,
+	}
+}
+
+func (rp *Reprovider) Run(ctx context.Context) {
+	after := time.After(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-after:
+			rp.reprovide(ctx)
+			after = time.After(time.Hour * 12)
```

Comment:
12 hours is a long time. we probably can't pick a good number without being able to understand how the whole net is behaving. since there is an initial reprovide at startup, this may be fine.